### DISCUSSION TITLE
Why does HttpClient dispatch it's response callbacks on the main thread?

### DISCUSSION THREAD

**Discussion by erawhctim** [braintree/braintree_android#1100] [General]
Not exactly a bug, per se, as it seems like very intentional behavior..... but it's caused some `NetworkOnMainThreadException`s that we've had to track down and debug through the SDK internals. 

Are the responses dispatched on main just for convenience? Or some other reason?

**Comment by sshropshire** [comment]
Hey @erawhctim thinking back, it was mostly for convenience so merchants wouldn't have to worry about switching back to the main thread. I do feel that the current implementation is overly opinionated and we may want to alter this behavior in the future. We are in the process of migrating to Kotlin. Moving to coroutines would allow us to cleanly delegate the responsibility of thread scheduling to the caller.

At the same time I am surprised that a `NetworkOnMainThreadException` is thrown. We do have a [ThreadScheduler](https://github.com/braintree/braintree_android/blob/2fea059ee908fc25b95e7e1deb344e16545b5679/SharedUtils/src/main/java/com/braintreepayments/api/sharedutils/ThreadScheduler.java#L11) class that allows HttpClient to execute the actual networking call [on a background thread](https://github.com/braintree/braintree_android/blob/f13626a796ae46b21fe41d62b0092c7f73d3c036/SharedUtils/src/main/java/com/braintreepayments/api/sharedutils/HttpClient.java#L70).

I am concerned now that the `ExecutorService` could possibly dispatch to main in some circumstances though. The [documentation](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool) doesn't explicitly rule submission of a `Runnable` to main. Do you have more details on when this occurs? Or if it's frequent?

**Reply by erawhctim** [reply]
> Moving to coroutines would allow us to cleanly delegate the responsibility of thread scheduling to the caller.

Totally agree! I think that rewrite is a great place to try and enhance this behavior 💯 

The `NetworkOnMainThreadException` we ran into seems to stem from [this bit here](https://github.com/braintree/braintree_android/blob/main/SharedUtils/src/main/java/com/braintreepayments/api/sharedutils/HttpClient.java#L70-L73) in the `HttpClient`.

**Reply by sshropshire** [reply]
Yeah the `ExecutorService` we create [here](https://github.com/braintree/braintree_android/blob/2fea059ee908fc25b95e7e1deb344e16545b5679/SharedUtils/src/main/java/com/braintreepayments/api/sharedutils/ThreadScheduler.java#L17) seems like it could include the main thread in it's pool somehow. It's probably unlikely and [the docs](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool--) don't explicitly call out that it is possible, but they also don't say it's impossible.
